### ISSUE TITLE
Weird multipart/chunked parsing bug

### ISSUE THREAD

**Issue by nigredo-tori** [http4s/http4s#2610]
I don't know how I can shrink this further... This seems to be some race/interplay between multipart and chunked encoding parsing, but I'm not sure.

Test case:
```scala
import cats.effect._
import cats.implicits._
import fs2.{Chunk, Stream}
import java.util.concurrent.Executors
import org.http4s._
import org.http4s.client.blaze.BlazeClientBuilder
import org.http4s.dsl.Http4sDsl
import org.http4s.implicits._
import org.http4s.multipart._
import org.http4s.server.blaze.BlazeServerBuilder
import org.specs2.execute.Result
import scala.concurrent.ExecutionContext
import scala.concurrent.duration._

class WeirdBugSpec extends org.specs2.mutable.Specification {

  private val log = org.slf4j.LoggerFactory.getLogger(getClass)

  val cpuEC = ExecutionContext.fromExecutorService(
    Executors.newWorkStealingPool
  )

  implicit val cs = IO.contextShift(cpuEC)
  implicit val timer = IO.timer(cpuEC)

  "Weird bug test case" >> {

    val numRequests = 100
    val responseDelay = 300.milliseconds

    val expectedResponse = Chunk.bytes("expected response".getBytes)

    val routes: HttpRoutes[IO] = {
      val dsl = Http4sDsl[IO]
      import dsl._

      HttpRoutes.of[IO] {
        case req @ POST -> Root =>
          req.decode[Multipart[IO]] { mp =>
            timer.sleep(responseDelay) *>
              // Has to be a `Stream` to trigger chunked encoding.
              Ok(Stream.chunk[IO, Byte](expectedResponse))
          }
      }
    }

    val mkServer = {
      BlazeServerBuilder[IO]
        .bindHttp(8080, "0.0.0.0")
        .withHttpApp(routes.orNotFound)
        .resource
    }

    val mkClient = BlazeClientBuilder[IO](cpuEC).resource

    val body = Multipart(
      Vector(
        Part.fileData[IO]("foo", "foo", Stream.emits("This is foo".getBytes)),
      ),
      // fewer moving parts this way
      boundary = Boundary("MwjRT0CXPTcG0AVzl/10g_uhN3Z7zk:IIJXJCvh/6")
    )

    val req =
      Request[IO](
        method = Method.POST,
        uri = Uri.unsafeFromString("http://localhost:8080")
      ).withEntity(body)
        // https://github.com/http4s/http4s/issues/146
        .withHeaders(body.headers)

    mkServer.use { _ =>
      mkClient.use { client =>
        val checkOnce = client.expect[Chunk[Byte]](req)
          .flatMap { result => IO {
            result must_=== expectedResponse
          }}.onError { case err => IO {
            log.error("Client error", err)
          }}

      (1 to numRequests).toVector
        .traverse(_ => checkOnce)
        .map(Result.forall(_)(_.toResult))
      }
    }.unsafeRunSync
  }
}
```
This fails with an error:
```
[blaze-selector-1] INFO  o.h.b.c.n.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:53504 
[scala-execution-context-global-1794] INFO  o.h.s.b.Http1ServerStage$$anon$1 - HTTP body not read to completion. Dropping connection. 
[ForkJoinPool-38-worker-4] ERROR WeirdBugSpec - Client error 
org.http4s.InvalidBodyException: Received premature EOF.
```
The exchanges between the server and the client look like shown below, until the server just decides to reset after sending the headers.
```http
POST / HTTP/1.1
Host: localhost:8080
Content-Type: multipart/form-data; boundary="MwjRT0CXPTcG0AVzl/10g_uhN3Z7zk:IIJXJCvh/6"
Accept: */*
User-Agent: http4s-blaze/0.20.1
Transfer-Encoding: chunked

2d
--MwjRT0CXPTcG0AVzl/10g_uhN3Z7zk:IIJXJCvh/6

5f
Content-Disposition: form-data; name="foo"; filename="foo"
Content-Transfer-Encoding: binary

2


b
This is foo
2f

--MwjRT0CXPTcG0AVzl/10g_uhN3Z7zk:IIJXJCvh/6--
0

HTTP/1.1 200 OK
Date: Thu, 30 May 2019 11:51:25 GMT
Transfer-Encoding: chunked

11
expected response
0
```
Environment: http4s 0.20.1, Scala 2.12.8, OpenJDK 1.8.0_212 x64, Ubuntu 18.04.2 LTS.

**Comment by nigredo-tori** [comment]
I've dug a bit into this. Judging by the symptoms, it seems like there are a few shady things going on here:
1. `MultipartParser` stops consuming (that is, cancels the underlying stream) once it finds the last boundary. The epilogue is discarded but not drained. When we're not using `chunked` encoding, everything is fine - since no one includes epilogue anyway, at this point we have consumed the whole message body. However with `chunked` we can end in the position where we have pulled all the encoded data, but not the terminating zero-length "chunk". This is a problem because...
1. The `chunked` parser seems to assume that downstream will keep pulling until error, or until it drains the whole encrypted body. Since `MultipartParser` has its own termination logic, we can end up in the situation described above... Which *should* be handled semi-properly - the server should close the connection and open a new one. Except...
1. There is some race between connection cleanup (which makes sure the whole message body was consumed), and writing to the stream. I have no other way to explain the connection being dropped in the middle of writing a message (between the headers and the body).

I'll send a PR for the `MultipartParser` to drain the epilogue - which should be enough to solve my immediate problem. However, the second and the third issues should still be addressed at some point.
